analyzer/src/main/kotlin/managers/Bower.kt

```diff
@@ -19,20 +19,237 @@
+
+        private fun extractVcsInfo(node: JsonNode) =
+                VcsInfo(
+                        type = extractRepositoryType(node),
+                        url = extractRepositoryUrl(node),
+                        revision = extractRevision(node)
+                )
+
+        private fun extractDeclaredLicenses(node: JsonNode): SortedSet<String> {
+            val result = mutableSetOf<String>()
+
+            val license = node["pkgMeta"]["license"].textValueOrEmpty()
+            if (!license.isEmpty()) {
+                result.add(license)
+            }
+
+            return result.toSortedSet()
+        }
+
+        private fun extractPackage(node: JsonNode ): Package {
+            val vcsInfo = extractVcsInfo(node)
+            return Package(
+                    id = extractPackageId(node),
+                    declaredLicenses = extractDeclaredLicenses(node),
+                    description = node["pkgMeta"]["description"].textValueOrEmpty(),
+                    homepageUrl = node["pkgMeta"]["homepage"].textValueOrEmpty(),
+                    binaryArtifact = RemoteArtifact.EMPTY,
+                    sourceArtifact = RemoteArtifact.EMPTY, // TODO: implement me!
+                    vcs = vcsInfo,
+                    vcsProcessed = vcsInfo.normalize()
+            )
+        }
+
+        private fun getDependencyNodes(node: JsonNode): Sequence<JsonNode> =
+                node["dependencies"].fieldsOrEmpty().asSequence().map { it.value }
+
+        private fun extractPackages(node: JsonNode): Map<String, Package> {
+            val result = mutableMapOf<String, Package>()
+
+            val stack = Stack<JsonNode>()
+            stack.addAll(getDependencyNodes(node))
+
+            while (!stack.empty()) {
+                val currentNode = stack.pop()
+                val pkg = extractPackage(currentNode)
+                result.put("${pkg.id.name}:${pkg.id.version}", pkg)
+
+                stack.addAll(getDependencyNodes(currentNode))
+            }
+
+            return result
+        }
+
+        private fun extractDependencyTree(node: JsonNode, scopeName: String): SortedSet<PackageReference> {
+            val result = mutableSetOf<PackageReference>()
+
+            node["pkgMeta"][scopeName].fieldNamesOrEmpty().forEach {
+                val childNode = node["dependencies"][it]
+                val childScope = SCOPE_NAME_DEPENDENCIES
+                val childDependencies = extractDependencyTree(childNode, childScope)
+                val packageRefence = PackageReference(
+                        id = extractPackageId(childNode),
+                        dependencies = childDependencies
+                )
+                result.add(packageRefence)
+            }
+
+            return result.toSortedSet()
+        }
+    }
+
+    private val stashedDirectories = HashMap<File, File>()
+
+    override fun toString(): String = PROVIDER_NAME
```

Comment:
Is it necessary to provide the return type here?

Reply:
No, it's not 😉